### PR TITLE
Fix multiple SCM Integration Issues

### LINKED ISSUE #5179: Add a button to delete the SCM configuration completely
**Describe the solution you'd like**
A new button on the SCM configuration screen to delete completely the scm-import.properties/scm-export.properties configuration files.

**Describe alternatives you've considered**
You can disable the SCM configuration and it has the same effect but it would be nice to have the capability to completely remove it.

**Additional context**
When you are testing a project or setting up a wrong project, it is desirable to be able to completely eliminate the information so that there are no further doubts as to why this configuration is unused

A new button beside the Clean button, to perform the same cleaning as the clean button and delete the configuration of the storage or the local files depending on the server configuration.
This button should only appears if the configuration is disabled and the user has permission to modify the configuration.

image

### LINKED ISSUE #5815: SCM Import shows 'N tracked files were deleted'
**Describe the bug**
When using the SCM plugin to sync with Git, after configuring a project and importing the jobs, if the SCM settings are changed (e.g. different Git branch to pull from), then after another import from the new Git config, the interface is always showing a message about deleted tracked files (`N tracked file(s) were deleted`) and there doesn't seem to be anything to do about it. I don't see the old files on the server.

**My Rundeck detail**

* Rundeck version: `3.1.2.20190927`
* install type: rpm
* OS Name/version: Linux/CentOS 7
* DB Type/version: MySQL

**To Reproduce**
1. Configure Rundeck to import the jobs from Git (SCM plugin)
2. Import the jobs from Git
3. Edit the SCM configuration to pull from another repo or branch, which would involve removing some of the jobs that were tracked by the previous configuration
4. Import the jobs after the config change
5. Observe that Rundeck now shows the SCM message `N tracked file(s) were deleted`

**Expected behavior**
After changing the SCM configuration, the changes should be imported and the Rundeck interface shouldn't show any message.

**Desktop (please complete the following information):**
 - OS: Windows 10
 - Browser: Chrome
 - Version: 80

### PR DESCRIPTION
**Pull request by ltamaster**
**Is this a bugfix, or an enhancement? Please describe.**
SCM Import/Export bug fixes and enhancements:

* Fix error 500 in the manual pull from remote repository.
* Storage was getting retrieved from cache, this was causing inconsistency issues when in cluster mode
* It validates that the SCM base directory configuration is not shared between any SCM configuration, either on the same project or with other projects
* Now it checks if the jobs being tracked were actually tracked for the configured repository, if not it will be removed from the validation to choose what message to show in the jobs page.

Enhancement: 
* User is now able to delete the SCM configuration once disabled
* Detects when local repo has a different version than jobs definition. Sync local repo version with the job definition.

Issues references: 

https://github.com/rundeckpro/rundeckpro/issues/1520
https://github.com/rundeckpro/rundeckpro/issues/1521
https://github.com/rundeckpro/rundeckpro/issues/1522
https://github.com/rundeckpro/rundeckpro/issues/1523
fixes #5179 
https://github.com/rundeckpro/rundeckpro/issues/1560
https://github.com/rundeckpro/rundeckpro/issues/1559
fixes #5815

### CODE REVIEW

File: rundeckapp/grails-app/services/rundeck/services/ScmService.groovy [hunk 1]
```diff
@@ -557,6 +570,42 @@ class ScmService {
...
+                    for(String allowedIntegration : INTEGRATIONS){
+                        if((project.equals(projectName) && !allowedIntegration.equals(integration)) || !project.equals(projectName)){
+                            ScmPluginConfigData scmPluginConfig = loadScmConfig(projectName, allowedIntegration)
+                            if(scmPluginConfig && scmPluginConfig.type){
+                                def scmPluginIntegration = loadPluginWithConfig(allowedIntegration, context, scmPluginConfig.type, scmPluginConfig.getConfig(), false)
+                                if(currentScmPluginIntegration.provider.getBaseDirectoryPropertyValue() &&
+                                        scmPluginIntegration.provider.getBaseDirectoryPropertyValue() &&
+                                        currentScmPluginIntegration.provider.getBaseDirectoryPropertyValue().trim().equals(scmPluginIntegration.provider.getBaseDirectoryPropertyValue().trim())){
+                                    if(integration.equals(EXPORT)){
+                                        loadedExportPlugins.remove(project)
```
Comment (gschueler) [inline rundeckapp/grails-app/services/rundeck/services/ScmService.groovy:595]: why is it removing the loadedExportPlugin here?

**Comment by ltamaster** [comment]
@Mergifyio rebase

File: rundeckapp/grails-app/services/rundeck/services/ScmService.groovy [hunk 2]
```diff
@@ -557,6 +570,42 @@ class ScmService {
...
+     * @param integration
+     * @param context
+     * @throws ScmPluginException
+     */
+    void validateIntegrationDirectory(project, integration,context) throws ScmPluginException {
+        ScmPluginConfigData currentPluginConfig = loadScmConfig(project, integration)
+        if(currentPluginConfig){
+            def currentScmPluginIntegration = loadPluginWithConfig(integration, context, currentPluginConfig.type, currentPluginConfig.getConfig(), false)
+            if(currentScmPluginIntegration && currentScmPluginIntegration.provider && currentScmPluginIntegration.provider.getBaseDirectoryPropertyValue()){
+                for(String projectName : frameworkService.getRundeckFramework().getFrameworkProjectMgr().listFrameworkProjectNames()){
```
Comment (gschueler) [inline rundeckapp/grails-app/services/rundeck/services/ScmService.groovy:585]: this is going to be very inefficient with a large number of projects.  can it just iterate on existing loadedPlugins?

**Comment by ltamaster** [comment]
@Mergifyio rebase

**Review by gschueler — CHANGES_REQUESTED** [review]
I think `validateIntegrationDirectory` semantics should be moved into the git plugin.  it already does validation to check a unique project is used, it could also verify the import/export is unique for the basedir.

**Review by gschueler — CHANGES_REQUESTED** [review]
the method `loadPluginWithConfig` (and `loadExportPluginWithConfig` and `loadImportPluginWithConfig`) return a `CloseableProvider`.  This is provides a Closeable interface, and must be closed after it is used.  The normal flow for a plugin will initialize it at startup or first configuration and then close it when it is done.  There are now some other calls to `loadPluginWithConfig` which aren't correctly closing the result after they are done

**Comment by gschueler** [comment]
I fixed one case where loadPluginWithConfig result was not being closed

**Comment by gschueler** [comment]
testing using a local build and https://github.com/rundeckpro/scm-docker-demo

* fix #1523 confirmed for local dir
* for #1520 seems ok
* for #1521 weird behavior, and I could not verify job renaming works, because I cannot rename a job.
    * Actually i cannot edit a job at all via Rundeck GUI.  No changes in the job edit form are being applied
    * deleting a job works
    * creating a job works
    * no other changes take effect

**Comment by gschueler** [comment]
* 1522: after cleaning up the test git repo to a clean state, i confirmed this is working as expected.  After pushing a new job to export repo, the job state is "?" and says "Export Status: Clean, not tracked for SCM Import"

**Comment by gschueler** [comment]
confirmed for:

fixes #5179
rundeckpro/rundeckpro#1560
rundeckpro/rundeckpro#1559

**Comment by gschueler** [comment]
#1585 seems fixed

**Comment by gschueler** [comment]
investigating the issue with being unable to modify a job, but not sure if it is related to scm changes or not

**Review by gschueler — APPROVED** [review]

**Comment by gschueler** [comment]
confirmed, the job rename issue is present in enterprise snapshot, I believe

**Comment by gschueler** [comment]
confirmed #1521 appears fixed

**Pull request merged**